File: apps/api/adapters/driving/http/Wire.kt
// JSON wire types for the HTTP adapter. Kept separate from the
// driving-port types (BatchInput / EventInput) so JSON / Jackson
// concerns don't leak into the inner hexagon.
package dev.mtrace.api.adapters.driving.http

import com.fasterxml.jackson.annotation.JsonProperty
import dev.mtrace.api.hexagon.port.driving.BatchInput
import dev.mtrace.api.hexagon.port.driving.EventInput
import dev.mtrace.api.hexagon.port.driving.SDKInput

data class WireBatch(
    @JsonProperty("schema_version") val schemaVersion: String = "",
    @JsonProperty("events") val events: List<WireEvent> = emptyList(),
)

data class WireEvent(
    @JsonProperty("event_name") val eventName: String = "",
    @JsonProperty("project_id") val projectId: String = "",
    @JsonProperty("session_id") val sessionId: String = "",
    @JsonProperty("client_timestamp") val clientTimestamp: String = "",
    @JsonProperty("sequence_number") val sequenceNumber: Long? = null,
    @JsonProperty("sdk") val sdk: WireSDK = WireSDK(),
)

data class WireSDK(
    @JsonProperty("name") val name: String = "",
    @JsonProperty("version") val version: String = "",
)

internal fun WireBatch.toBatchInput(authToken: String): BatchInput = BatchInput(
    schemaVersion = schemaVersion,
    authToken = authToken,
    events = events.map { it.toEventInput() },
)

internal fun WireEvent.toEventInput(): EventInput = EventInput(
    eventName = eventName,
    projectId = projectId,
    sessionId = sessionId,
    clientTimestamp = clientTimestamp,
    sequenceNumber = sequenceNumber,
    sdk = SDKInput(name = sdk.name, version = sdk.version),
)

File: apps/api/adapters/driving/http/wire.go
// Package http is the inbound HTTP adapter for the spike. It parses
// the JSON wire format into driving.BatchInput, calls the use case,
// and maps domain errors to HTTP status codes per
// spec/backend-api-contract.md §5.
package http

import "github.com/pt9912/m-trace/apps/api/hexagon/port/driving"

// wireEvent is the JSON shape of a single event in the request body.
// The use case never sees these tags; mapping happens in
// toEventInputs.
type wireEvent struct {
	EventName       string         `json:"event_name"`
	ProjectID       string         `json:"project_id"`
	SessionID       string         `json:"session_id"`
	ClientTimestamp string         `json:"client_timestamp"`
	SequenceNumber  *int64         `json:"sequence_number,omitempty"`
	SDK             wireSDK        `json:"sdk"`
	Meta            map[string]any `json:"meta,omitempty"`
}

type wireSDK struct {
	Name    string `json:"name"`
	Version string `json:"version"`
}

// wireBoundary spiegelt einen Eintrag aus dem optionalen
// `session_boundaries[]`-Block (API-Kontrakt §3.4). Use-Case validiert
// alle Felder; HTTP-Adapter dekodiert nur das Wire-Format.
type wireBoundary struct {
	Kind            string `json:"kind"`
	ProjectID       string `json:"project_id"`
	SessionID       string `json:"session_id"`
	NetworkKind     string `json:"network_kind"`
	Adapter         string `json:"adapter"`
	Reason          string `json:"reason"`
	ClientTimestamp string `json:"client_timestamp"`
}

type wireBatch struct {
	SchemaVersion     string         `json:"schema_version"`
	Events            []wireEvent    `json:"events"`
	SessionBoundaries []wireBoundary `json:"session_boundaries,omitempty"`
}

func toBoundaryInputs(boundaries []wireBoundary) []driving.BoundaryInput {
	if len(boundaries) == 0 {
		return nil
	}
	out := make([]driving.BoundaryInput, len(boundaries))
	for i, b := range boundaries {
		out[i] = driving.BoundaryInput{
			Kind:            b.Kind,
			ProjectID:       b.ProjectID,
			SessionID:       b.SessionID,
			NetworkKind:     b.NetworkKind,
			Adapter:         b.Adapter,
			Reason:          b.Reason,
			ClientTimestamp: b.ClientTimestamp,
		}
	}
	return out
}

func toEventInputs(events []wireEvent) []driving.EventInput {
	out := make([]driving.EventInput, len(events))
	for i, e := range events {
		out[i] = driving.EventInput{
			EventName:       e.EventName,
			ProjectID:       e.ProjectID,
			SessionID:       e.SessionID,
			ClientTimestamp: e.ClientTimestamp,
			SequenceNumber:  e.SequenceNumber,
			SDK: driving.SDKInput{
				Name:    e.SDK.Name,
				Version: e.SDK.Version,
			},
			Meta: copyMeta(e.Meta),
		}
	}
	return out
}

func copyMeta(in map[string]any) map[string]any {
	if len(in) == 0 {
		return nil
	}
	out := make(map[string]any, len(in))
	for k, v := range in {
		out[k] = v
	}
	return out
}
